### PR TITLE
Ticket #26 -  Add Search results page filters component

### LINKED ISSUE #26: Add Search results page filters component
## Description
Add component for filters

Notes: 
- Filters should include options from figma
- Options can be maintained as a map constant on the client side for now
- nice to have: Selecting `Show results` adds the results to query params
- nice to have: Parse query params to pre check any filters   

image

image

### PR DESCRIPTION
**Pull request by JennaXiao3**
## Description
Fixes #26 

#### What is the purpose of your change?
- to add a filter bar to search results page

#### What areas of the application does this change impact? (changes API, folder structure, etc)
- for now, added a sample object containing all the info of the filters in SearchResultsPage

#### What features/fixes are deferred/not handled in this PR?
- fetching data from backend

Notes: might be a bit confusing since there is a lot going on - let me know if there are any questions

## Checklist
- [ ] Adds unit tests
- [ ] Has a work item linked to the PR

### CODE REVIEW

File: client/src/components/Filter.jsx
Comment (jwu910) [inline client/src/components/Filter.jsx:183]: Just out of curiosity, do you have a preference over grid vs flex? Personally i don't find too many instances where grid is helpful for much smaller components like this, and usually more helpful forcing columns on larger layouts.

Not saying this is wrong, just Grid adds so many elements vs a few styling options for flex containers and flex items.

File: client/src/components/Filter.jsx
Comment (jwu910) [inline client/src/components/Filter.jsx:88]: If we turn on react dev tool's update painting, clicking on one checkbox looks like we're repainting a lot of the popover, meaning the whole component is updating. If we minimize the state updates, we might be able to clean this up as more of a performance change and keeping the logic a little more succinct.

File: client/src/components/Navbar.jsx
Comment (jwu910) [inline client/src/components/Navbar.jsx:31]: is there a need for `target` and `rel` here?

File: client/src/pages/SearchResultsPage.jsx
Comment (jwu910) [inline client/src/pages/SearchResultsPage.jsx:135]: is this just mock data? if so, we should probably pull this out into a constant and create a method to mock getting the data and supplying the state here

File: client/src/components/Filter.jsx
Comment (jwu910) [inline client/src/components/Filter.jsx:71]: maybe just a nit pick but this organization of the inline `if` conditions with an `else` following w/o curlies feels confusing to me. but see other comments about button names first

File: client/src/components/Filter.jsx
Comment (jwu910) [inline client/src/components/Filter.jsx:36]: left some comments about state here, lmk if you need to discuss anything more, we have quite a bit of state that might be more than we need here

**Review by jwu910 — CHANGES_REQUESTED** [review]
nice job, filters are always confusing and can get complex so quick. Left a few comments and questions, lmk if you need to review any of them

File: client/src/components/Filter.jsx
Comment (JennaXiao3) [reply client/src/components/Filter.jsx:88]: woww how do you do that

File: client/src/components/Filter.jsx
Comment (JennaXiao3) [reply client/src/components/Filter.jsx:221]: if I try to use a function the page will not display due to too many re-renders

Comment (jwu910) [reply client/src/components/Filter.jsx:221]: Might not necessarily rule this out though. Next question would be why are there so many rerenders?

This would be dependent on `status` and might be a good lead to asking if that's changing too frequently

File: client/src/components/Filter.jsx
Comment (jwu910) [reply client/src/components/Filter.jsx:88]: image

https://react.dev/learn/react-developer-tools

browser extension!

File: client/src/components/Filter.jsx
Comment (JennaXiao3) [reply client/src/components/Filter.jsx:221]: yeah...I'm not sure why there are so many rerenders...instead i've created an array containing the labels since 0,1,2 from buttonStatus can represent the index of the array

File: client/src/components/Filter.jsx
Comment (JennaXiao3) [reply client/src/components/Filter.jsx:183]: ngl I completely forgot about the flex attribute...once I started MUI I kind of replaced flex with Grid...would you like me to use flex here instead to shorten the code?

File: client/src/pages/SearchResultsPage.jsx
Comment (JennaXiao3) [reply client/src/pages/SearchResultsPage.jsx]: can you elaborate?

File: client/src/components/Filter.jsx
Comment (JennaXiao3) [reply client/src/components/Filter.jsx]: i've reduced the number of states but it still seems to look the same...

File: client/src/components/Filter.jsx
Comment (jwu910) [reply client/src/components/Filter.jsx:159]: i think up to you, i dont see a problem with flex in this situation, and if grid is working, thats fine too. Grid just creates the need for more elements to organize spaces so small, and i think columns becomes tough to work with at that point. But also if we're not using columns, maybe we just use flex

File: client/src/pages/SearchResultsPage.jsx
Comment (jwu910) [inline client/src/pages/SearchResultsPage.jsx:142]: One thing we want to consider maybe for later too, is persisting the filters for future use. If a user selects 5 different filter options, and then refreshes their browser, they'll lose the time the spent selecting those options to filter their results.

Are there any solutions that you can think of that would solve that?

File: client/src/pages/SearchResultsPage.jsx
Comment (jwu910) [reply client/src/pages/SearchResultsPage.jsx]: Can do something like 

```js
const MOCK_FILTER_DATA = {...}

const [filterData, setFilterData] = useState(MOCK_FILTER_DATA)
```

however, there will be logic to actually GET the available filters from the backend first right? We will have to create an empty object according to the available options.

pseudo logic might be something like

// fetch internships
// reduce internships to get dates posted, semesters, education, and company
// take this data to create keys in this object
// take this data to create counts in this object

we can revisit this once we get the endpoint though

but we can talk to @a-camarillo and figure out how the endpoint response would look like or expect to look like so we can start working on that logic as well, so we can write out the fetch logic still and just return a mock list of internships so we can extract this infromation

File: client/src/theme.js
Comment (jwu910) [inline client/src/theme.js:97]: are these theme colors we should pull from the palette?

Comment (JennaXiao3) [reply client/src/theme.js:97]: I've been trying to figure this out. The problem is that the palette is defined in theme and so is this part of the theme so it cannot refer to itself to get the colours if it is part of the definition of theme itself. Another thing we can do is pull this part out of the theme and define it elsewhere while referring to the theme but then I would have to redefine the same thing again to add more to it and I feel like it will get messy and disorganized

File: client/src/pages/SearchResultsPage.jsx
Comment (JennaXiao3) [reply client/src/pages/SearchResultsPage.jsx]: Okay yep that makes sense

File: client/src/theme.js
Comment (jwu910) [reply client/src/theme.js:97]: Oh yeah so we could also add another named color map to the theme file if we want, like a hash map

```js
const COLORS = {
  taro: '#123123',
  redBean: '#123123',
}
```

as well, and then in the theme definitions, we can extend that

```js
export const theme = createTheme({
  palette: {
    primary: {
      main: COLORS.taro
    }
  }
})
```

File: client/src/pages/SearchResultsPage.jsx
Comment (a-camarillo) [reply client/src/pages/SearchResultsPage.jsx]: Still brainstorming the endpoints for filtering, at the moment I'm not entirely sure if it's better design to have the filters be passed in as parameters to the jobs endpoint i.e. `GET /jobs?company_name=Google` or to have specific endpoints for the companies i.e. `GET /jobs/google`. 

Additionally there's some information that is being expected to be presented that I'm not entirely sure we can supply. Specifically, I know information related to date posted and semester isn't available on all company job posts. I suppose as a work around we can supply an N/A value so we're not eliminating the option from companies that do share the information.

Comment (JennaXiao3) [reply client/src/pages/SearchResultsPage.jsx]: @a-camarillo We can filter the information in the frontend so you don't have to worry about passing parameters to the endpoint.

For information that can't be shown, can you communicate that to the design team? They can maybe determine if it's nicer to just completely remove the field or putting in a placeholder like N/A like you mentioned

File: client/src/pages/SearchResultsPage.jsx
Comment (JennaXiao3) [reply client/src/pages/SearchResultsPage.jsx:142]: I've used local storage to solve this problem

Comment (jwu910) [reply client/src/pages/SearchResultsPage.jsx:155]: What if the user wants to share the filter with another student?

More just probing to think about some other possibilities, but I think local storage is a viable solution as well for certain use cases

File: client/src/theme.js
Comment (JennaXiao3) [reply client/src/theme.js:97]: should we make this another ticket or should I directly do it directly here? (rather than everyone making changes to their own ticket to adapt to this change we can make all the changes in a new ticket)

File: client/src/pages/SearchResultsPage.jsx
Comment (JennaXiao3) [reply client/src/pages/SearchResultsPage.jsx:155]: wdym by sharing filters? would there be a button to save the filter and share the link to someone else?
would you like me to keep the local storage for now?

Comment (michachan) [reply client/src/pages/SearchResultsPage.jsx:155]: I think the hint is to use query parameters to set the filters on render/useEffect, such that if anyone had that link, they'd be able to set the filters in that way

Comment (jwu910) [reply client/src/pages/SearchResultsPage.jsx:155]: What Michael said! If we store filters as query params, users could also bookmark different combinations of filters if they wanted

Comment (JennaXiao3) [reply client/src/pages/SearchResultsPage.jsx:155]: ohhh that's cool I don't think I've ever done that in this context before

File: client/src/theme.js
Comment (JennaXiao3) [reply client/src/theme.js:97]: Yeah, I think this would help with organization and matching with the design. One thing I'm worried about is that there are a lot of colours that are used in the figma and most of them are unnamed or used only once or twice. It might difficult to match a colour to a name in the theme.

What do you think?

File: client/src/pages/SearchResultsPage.jsx
Comment (JennaXiao3) [reply client/src/pages/SearchResultsPage.jsx:155]: Is this done using useSearchParams hook? Should we put this in a new ticket?

File: client/src/theme.js
Comment (JennaXiao3) [reply client/src/theme.js:97]: Yeah, I understand how this would benefit in terms of organization. One problem I notice is that there are a lot of colours, and most colors in the figma are unnamed and are only used once or twice. If we do it this way, we will be basically giving every colour two names...and most will not be repeated.

I just thought of another way we can try to do it that is similar to your method...maybe we can define all the colours in a separate object and then add it to the theme like so:

```
const COLORS = {
   brown: {
      ...
   },
   gray: {
      ...
   },
};

const theme = createTheme({
   palette: {
       ...COLORS
   },
   components: {
       MuiButton: {
           color: COLORS.brown[100],
       }
   },
});
```

Comment (jwu910) [reply client/src/theme.js:97]: If there are too many diff colors in the design, it might be an opportunity to ask design and potentially correct an unintentional color variation, if design wants 10 diff shades of brown though, we can do it, too

But never hurts to ask, designs aren't the final say, as devs I think it's important to pushback a little too, question when things seem weird

File: client/src/pages/SearchResultsPage.jsx
Comment (jwu910) [reply client/src/pages/SearchResultsPage.jsx:155]: We can push it to a new ticket, too, local storage probably okay, but another concern we'd want to consider or ask: do we know the user wants to save everything on refresh?

We're forcing users to manually select each filter and clearing them now. Might be a question for design on what that experience should be

**Comment by jwu910** [comment]
App preview isn't loading on my phone anymore, can't check console for errors, can you take a look before i review latest changes?

File: client/src/pages/SearchResultsPage.jsx
Comment (JennaXiao3) [reply client/src/pages/SearchResultsPage.jsx:155]: That's true...I've thought about that as well. I'll ask the design team. Can we push this to a new ticket while we wait for their response?

File: client/src/theme.js
Comment (JennaXiao3) [reply client/src/theme.js:97]: I'll ask the design 👍

**Comment by JennaXiao3** [comment]
> App preview isn't loading on my phone anymore, can't check console for errors, can you take a look before i review latest changes?

Yep of course!
Screen Shot 2023-04-23 at 10 26 28 AM

**Comment by JennaXiao3** [comment]
> App preview isn't loading on my phone anymore, can't check console for errors, can you take a look before i review latest changes?

Here is the console:
Screen Shot 2023-04-23 at 10 54 47 AM

**Comment by jwu910** [comment]
Is this all that's outputting on the page you worked on?

**Comment by JennaXiao3** [comment]
> Is this all that's outputting on the page you worked on?
> 
> I won't be free till maybe later tonight to take a look at my computer

Yes I believe so...it was on the search results page

**Comment by jwu910** [comment]
@michachan can you double check? I think we're okay squashing and merging if everything looks ok, don't get to a computer till later and want to try to keep things moving

**Comment by michachan** [comment]
checking

**Pull request merged**
